vendors/infineon/secure_elements/pkcs11/iot_pkcs11_trustx.c

```
@@ -0,0 +1,3298 @@
+            xClass = CKO_PRIVATE_KEY;
+            break;
+        case eAwsDevicePublicKey:
+        case eAwsTestPublicKey:
+            xClass = CKO_PUBLIC_KEY;
+            break;
+        case eAwsDeviceCertificate:
+        case eAwsTestCertificate:
+            xClass = CKO_CERTIFICATE;
+            break;
+        default:
+            xResult = CKR_DATA_INVALID;
+            break;
+        }
+    }
+
+    if( xResult == CKR_OK )
+    {
+        for( iAttrib = 0; iAttrib < ulCount && CKR_OK == xResult; iAttrib++ )
+        {
+            switch( pxTemplate[ iAttrib ].type )
+            {
+                case CKA_CLASS:
+
+                    if( pxTemplate[ iAttrib ].pValue == NULL )
+                    {
+                        pxTemplate[ iAttrib ].ulValueLen = sizeof( CK_OBJECT_CLASS );
+                    }
+                    else
+                    {
+                        if( pxTemplate[ iAttrib ].ulValueLen >= sizeof( CK_OBJECT_CLASS ) )
+                        {
+                            memcpy( pxTemplate[ iAttrib ].pValue, &xClass, sizeof( CK_OBJECT_CLASS ) );
+                        }
+                        else
+                        {
+                            xResult = CKR_BUFFER_TOO_SMALL;
+                        }
+                    }
+
+                    break;
+
+                case CKA_VALUE:
+
+                    if( xIsPrivate == CK_TRUE )
+                    {
+                        pxTemplate[ iAttrib ].ulValueLen = CK_UNAVAILABLE_INFORMATION;
+                        xResult = CKR_ATTRIBUTE_SENSITIVE;
+                    }
+                    else
+                    {
+                        if( pxTemplate[ iAttrib ].pValue == NULL )
+                        {
+                            pxTemplate[ iAttrib ].ulValueLen = ulLength;
+                        }
+                        else if( pxTemplate[ iAttrib ].ulValueLen < ulLength )
+                        {
+         
```

consider combining with else

---

Can you please clarify? The code looks correct to me (and it's at least consistent in the way it handles these "size queries").

---

yes - I just feel like having one if for a specific mechanism (CKM_RSA_PKCS) and then having an else catch-all is un-necessary.

Also, I don't think that CKM_RSA_PKCS should return a different error code than all the other unsupported mechanism.s

---

Agreed - can you please see the latest?